cmd/kubernetes-discovery/pkg/apiserver/handler_apis.go

```diff
@@ -108,18 +117,46 @@ func (r *apisHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
-func newDiscoveryAPIGroup(apiServices []*apiregistrationapi.APIService) *metav1.APIGroup {
+// newDiscoveryAPIGroup takes apiservices in a single group and returns a discovery compatible object.
+// if none of the services are available, it will return nil.
+func newDiscoveryAPIGroup(apiServices []*apiregistrationapi.APIService, serviceLister v1listers.ServiceLister, endpointsLister v1listers.EndpointsLister) *metav1.APIGroup {
 	apiServicesByGroup := apiregistrationapi.SortedByGroup(apiServices)[0]
 
-	discoveryGroup := &metav1.APIGroup{
-		Name: apiServicesByGroup[0].Spec.Group,
-		PreferredVersion: metav1.GroupVersionForDiscovery{
-			GroupVersion: apiServicesByGroup[0].Spec.Group + "/" + apiServicesByGroup[0].Spec.Version,
-			Version:      apiServicesByGroup[0].Spec.Version,
-		},
-	}
+	var discoveryGroup *metav1.APIGroup
 
 	for _, apiService := range apiServicesByGroup {
+		// skip any API services without actual services
+		if _, err := serviceLister.Services(apiService.Spec.Service.Namespace).Get(apiService.Spec.Service.Name); err != nil {
+			continue
+		}
+
+		hasActiveEndpoints := false
+		// skip any API services without endpoints
+		if endpoints, err := endpointsLister.Endpoints(apiService.Spec.Service.Namespace).Get(apiService.Spec.Service.Name); err != nil {
```

Comment:
> hoping that they endpoints are indexed by service name.

The informer and lister make it hard to do otherwise.  I'm ok with code like this